docs/getPaginatedResponseFromAggregate.md

```diff
@@ -0,0 +1,44 @@
+# getPaginatedResponseFromAggregate
+
+Returns a paginated response from a MongoDB aggregation pipeline.
+
+Usage example:
+
+```js
+import getPaginatedResponseFromAggregate from "@reactioncommerce/api-utils/getPaginatedResponseFromAggregate.js";
+
+// the MongoDB collection you want to run your aggregation on
+const collection = getYourMongoDBCollectionSomehow();
+
+const pipeline = [
+  {
+    // for example, here's a $lookup operator
+    $lookup: {
+      // some options
+    }
+  },
+  // but you can add,
+  // any,
+  // other,
+  // operators
+];
+
+const args = {
+  // connection arguments from your GraphQL query or mutation
+  // supported:
+  // offset: Int,
+  // last: Int,
+  // sortBy: String (the name of the MongoDB field you want to sort results by)
+  // sortOrder: Int (-1 for descending or 1 for ascending)
```

Comment:
I believe `sortOrder` is actually an enum string of either "asc" or "desc" at this point, and your function converts that to -1 or 1 using `getMongoSort` call.

You should also list `first`, `after`, and `before` args here, all `Int`. Maybe put this URL in the comment, too: https://facebook.github.io/relay/graphql/connections.htm#sec-Arguments

Reply:
Fixed `sortOrder` and added the URL. However, `first`, `after` and `before` aren't supported because of the nature of aggregates.

From what I understood, these filters aren't available through the MongoDB API for aggregates. We would need to transform the results to an array, and then apply them with JavaScript, which would work but would mean a noticeable performance hit.

Are these really a necessity in order to merge a first version of this function? It's true that maintaining 100% compatibility with the rest of the API would be great, but I'd like to know what your approach to that would be.

Reply:
Also, thanks for doing this code review on a Sunday! 👍

Reply:
If you mean that it only supports offset pagination and not cursor-based, that is fine. But offset pagination works only with `first`, not `last`. I even see this in the code:

```js
// offset and last cannot be used together
if (last) throw new Error("Request either `last` or `offset` but not both");
```

I'm not sure what "last" would mean in the context of offset pagination. That's for backwards pagination and offset can't be done backwards.

I do think it would be possible to support before/after/last for aggregates in much the same way we do for normal cursors -- by doing a few lookups and converting them to skip/limit numbers -- but that isn't critical for now. It seems like the code is _trying_ to support cursor-based pagination, though, so maybe you should explicitly throw errors if they try to pass any of those args?

Reply:
Sorry @aldeed for the confusion. I was tired when writing that.

I meant that `first` and `last` are supported, with `offset` when applicable. They work in the same as they do with `getPaginatedResponse`, only by using pipeline operators instead of cursor methods. `after` and `before` are the only connection args that aren't supported currently.

I'll add `first` to the docs and will add errors if `after` or `before` are used, to let people know while we work on implementing them.

Reply:
@aldeed Just pushed these changes, with some tests for the error thrown in case `after` or `before` is used. Let me know if that's good enough or if you'd like anything changed!